test/hexpm_web/controllers/api/release_controller_test.exs

```diff
@@ -49,6 +49,27 @@ defmodule HexpmWeb.API.ReleaseControllerTest do
       assert log.params["release"]["version"] == "1.0.0"
     end
 
+    test "create release and new package fails with invalid version", %{user: user} do
+      meta = %{
+        name: Fake.sequence(:package),
+        version: "1.0-prod",
+        description: "Domain-specific language."
+      }
+
+      conn =
+        build_conn()
+        |> put_req_header("content-type", "application/octet-stream")
+        |> put_req_header("authorization", key_for(user))
+        |> post("api/packages/#{meta.name}/releases", create_tar(meta, []))
+     
+      assert conn.status == 422
+      result = json_response(conn, 422)
+      assert result["message"] =~ "Validation error" 
+      assert result["errors"] =~ "is not semantic"
```

Comment:
If I keep your test changes but revert production code changes I get these 2 failures:

```
 oss/hexpm[pr-772]% mix test
................................................*.............................................................................................................................................................................................................................................................................................

  1) test POST /api/packages/:name/releases create release and new package fails with invalid version (HexpmWeb.API.ReleaseControllerTest)
     test/hexpm_web/controllers/api/release_controller_test.exs:52
     ** (FunctionClauseError) no function clause matching in Kernel.=~/2

     The following arguments were given to Kernel.=~/2:

         # 1
         %{"version" => "is invalid SemVer"}

         # 2
         "is not semantic"

     Attempted function clauses (showing 3 out of 3):

         def left =~ "" when is_binary(left)
         def left =~ right when is_binary(left) and is_binary(right)
         def left =~ right when is_binary(left)

     code: assert result["errors"] =~ "is not semantic"
     stacktrace:
       (elixir) lib/kernel.ex:1947: Kernel.=~/2
       test/hexpm_web/controllers/api/release_controller_test.exs:68: (test)

............

  2) test POST /api/packages/:name/releases update package fails when version is invalid (HexpmWeb.API.ReleaseControllerTest)
     test/hexpm_web/controllers/api/release_controller_test.exs:90
     ** (Ecto.Query.CastError) deps/ecto/lib/ecto/repo/queryable.ex:370: value `"1.0-dev"` in `where` cannot be cast to type Hexpm.Version in query:

     from r in Hexpm.Repository.Release,
       where: r.package_id == ^10243,
       where: r.version == ^"1.0-dev",
       select: r

     stacktrace:
       (elixir) lib/enum.ex:1934: Enum."-reduce/3-lists^foldl/2-0-"/3
       (elixir) lib/enum.ex:1426: Enum."-map_reduce/3-lists^mapfoldl/2-0-"/3
       (elixir) lib/enum.ex:1426: Enum."-map_reduce/3-lists^mapfoldl/2-0-"/3
       (elixir) lib/enum.ex:1934: Enum."-reduce/3-lists^foldl/2-0-"/3
       (ecto) lib/ecto/repo/queryable.ex:132: Ecto.Repo.Queryable.execute/4
       (ecto) lib/ecto/repo/queryable.ex:18: Ecto.Repo.Queryable.all/3
       (ecto) lib/ecto/repo/queryable.ex:66: Ecto.Repo.Queryable.one/3
       (hexpm) lib/hexpm/repository/releases.ex:169: Hexpm.Repository.Releases.create_release/5
       (hexpm) lib/hexpm/repository/releases.ex:46: Hexpm.Repository.Releases.publish/7
       (hexpm) lib/hexpm_web/controllers/api/release_controller.ex:118: HexpmWeb.API.ReleaseController.handle_tarball/5
       (hexpm) lib/hexpm_web/controllers/api/release_controller.ex:1: HexpmWeb.API.ReleaseController.action/2

.........................................................

Finished in 4.1 seconds
405 tests, 2 failures, 1 skipped
```

See the 1st failure. `result["errors"]` should be a map and as you can see the validation already works on master when creating the release.

FYI: the way it works is we use custom `Hexpm.Version` type for the `version` field and we have code in controller helpers that transforms it's error message to `"is invalid SemVer"`.

There's still a bug on updating the release exposed by your 2nd test, good catch!

Reply:
aye, so let's revert to "is invalid SemVer" then, missed that.